llvm/lib/Analysis/MemoryBuiltins.cpp

```diff
@@ -686,10 +686,23 @@ ObjectSizeOffsetVisitor::ObjectSizeOffsetVisitor(const DataLayout &DL,
 
 SizeOffsetAPInt ObjectSizeOffsetVisitor::compute(Value *V) {
   InstructionsVisited = 0;
-  return computeImpl(V);
+  OffsetSpan Span = computeImpl(V);
+
+  // In ExactSizeFromOffset mode, we don't care about the Before Field, so allow
+  // us to overwrite it if needs be.
+  if (Span.knownAfter() && !Span.knownBefore() &&
+      Options.EvalMode == ObjectSizeOpts::Mode::ExactSizeFromOffset)
+    Span.Before = APInt::getZero(Span.After.getBitWidth());
+
+  if (!Span.bothKnown())
+    return {};
+  if (Span.Before.isNegative() || Span.After.isNegative())
+    return {};
```

Comment:
Do we need this bit? If this condition is true, we know we have an out of bounds pointer, yet we return that we know nothing about the object's size. I believe if we take this out, the below `return {Span.Before + Span.After, Span.Before};` already handles it correctly and indicates to the caller that we know we have an out of bounds pointer.

A test case for this is

```c++
#include <stdio.h>
int x;
int main() {
  int array[4];
  int *ptr = array + 8;
  printf("%zu\n", __builtin_object_size(ptr, 0));
}
```
Any output is valid, but an output of `0` is best. Previously, we used to print `0`. With your PR, we print `18446744073709551615`. If we take these two lines out, we pass all existing tests, and print `0` again for this new test.

Reply:
Indeed. I've just forced size and offset to Zero in that case to make it clear we 're doing something wrong

Reply:
That may be right at the C++ level, but I think it's wrong at the LLVM IR level. At the LLVM IR level, we may generate `getelementptr` without the `inbounds` attribute, do pointer arithmetic that takes us outside of an object, and then do more pointer arithmetic to take us back in, think the equivalent of `(array + 8) - 6`. If we do nothing special, we will return that `array + 8` is 32 bytes after the start of an object of size 16. We can then compute that `(array + 8) - 6` is 8 bytes into an object of size 16, and allows 8 bytes to be accessed. If we instead return that `array + 8` is zero bytes after the start of an object of size zero, `(array + 8) - 6` would be considered out of bounds.

It may take some effort to come up with a concrete test case demonstrating this though.

Reply:
As you proposed at some point, I just removed the check and checked it was still passing the (now well enhanced) test suite, it's indeed the case. Way to go!